doc/api/n-api.md

```diff
@@ -890,6 +890,53 @@ If still valid, this API returns the `napi_value` representing the
 JavaScript Object associated with the `napi_ref`. Otherise, result
 will be NULL.
 
+### Cleanup on exit of the current Node.js instance
+
+While a Node.js typically releases all its resources when exiting, embedders
+of Node.js, or future Worker support, may require addons to register
+clean-up hooks that will be run once the current Node.js instance exits.
+
+N-API provides functions for registering and un-registering such callbacks.
+When those callbacks are run, all resources that are being held by the addon
+should be freed up.
+
+#### napi_add_env_cleanup_hook
+<!-- YAML
+added: REPLACEME
+-->
+```C
+NODE_EXTERN napi_status napi_add_env_cleanup_hook(napi_env env,
+                                                  void (*fun)(void* arg),
+                                                  void* arg);
+```
+
+Registers `fun` as a function to be run with the `arg` parameter once the
+current Node.js environment exits.
+
+A function can safely be specified multiple times with different
```

Comment:
one question: if I call `napi_add_env_cleanup_hook` multiple times with the same function and arg, how can I just remove one of them when calling `napi_remove_env_cleanup_hook`?

Reply:
@yhwang You can’t do it, right now it would crash with that – I’ve clarified this in the documentation.

The reason is that you basically never want that. If you pass the same arg + fun combination twice in a meaningful way, that means that you rely on global state; in that case, your code will probably not work *anyway* in the scenarios that this is meant to support.

(Another reason is the suspicion – not benchmarked – that a `std::unordered_set` is more performant than a `std::unordered_multiset`.)

Reply:
> If you pass the same arg + fun combination twice in a meaningful way, that means that you rely on global state

hmm, the `arg` could be a refcount object and decrease its counter in each call of the callback function :-p

I saw you said `with different arg values` now. Good catch!

Reply:
> > If you pass the same arg + fun combination twice in a meaningful way, that means that you rely on global state

> hmm, the `arg` could be a refcount object and decrease its counter in each call of the callback function :-p

Yeah, you could still get that kind of behaviour by creating pointers with an additional layer of indirection if you really want to and using those as unique keyes. But like I said, probably not something one really wants to do :)